Fix caml_make_float_vect

damiendoligez opened the pull request:
This PR fixes bugs revealed by the test file of #1936.

When OCaml is configured with `-no-flat-float-array`, the `Array.create_float` primitive is completely wrong:

Before this patch:
```
ocaml
        OCaml version 4.08.0+dev0-2018-04-09

# Array.create_float (-1);;
Out of memory during evaluation.
# Array.create_float 3;;
Segmentation fault: 11
```

After this patch:
```
ocaml
        OCaml version 4.08.0+dev0-2018-04-09

# Array.create_float (-1);;
Exception: Invalid_argument "Array.create_float".
# Array.create_float 3;;
- : float array = [|0.; 0.; 0.|]
```

cc @gasche

damiendoligez committed:
When configured with -no-flat-float-array, the primitive caml_make_float_vect
(known as Array.create_float in OCaml) is completely broken:
1. it fails to raise Invalid_argument when the size is negative or too large.
2. it fails to initialize its result, which immediately leads to segfault.

damiendoligez committed:
fix message for invalid argument

damiendoligez committed:
add Changes entry

yallop commented on runtime/array.c:
```diff
@@ -290,7 +290,18 @@ CAMLprim value caml_make_float_vect(value len)
+  if (wosize > Max_wosize) caml_invalid_argument ("Array.create_float");
+  result = caml_alloc (Long_val (len), 0);
+  for (i = 0; i < wosize; i++){
+    x = caml_alloc_small (Double_wosize, Double_tag);
```
If I'm not mistaken, this function could be made quite a bit faster by moving this allocation out of the loop (i.e. by initializing every element of the array with the same value).

damiendoligez replied:
Yes, I realized that on my way home... I'll change this PR soon.

damiendoligez committed:
#2183: optimize and simplify

gasche commented on runtime/array.c:
```diff
@@ -346,6 +335,20 @@ CAMLprim value caml_make_vect(value len, value init)
+#else
+  /* allocate the float in the major heap directly to avoid triggering
+     a minor GC in [caml_make_vect]. */
+  value x = caml_alloc_shr (Double_wosize, Double_tag);
```
With the proposed implementation, we can drain the major heap by repeatedly allocating and discarding empty (or very small) float arrays in a tight loop. Wouldn't it be better to use a local root in the minor heap, or a global float value instead?

damiendoligez replied:
How many programs are going to allocate lots of small *uninitialized* float arrays?

A local root in the minor heap would trigger a minor collection every time the new float array is too large for the minor heap. I think that's more likely to be a problem.

The global float value is a better solution. I'll try to implement that.

damiendoligez replied:
Done.

gasche reviewed (commented):
I'm a bit worried by the amount of code duplication in the `array.c` file. Why cannot we share more logic, with the generic version calling either the `floatarray` functions or the `float_array` functions depending on whether FLAT_FLOAT_ARRAY is set?

damiendoligez commented:
>Why cannot we share more logic

This should probably be done in another PR.

damiendoligez committed:
In -no-flat-float-array mode, allocate the uninitialized float value (used
for Array.create_float) only once and reuse it for all new "uninitialized"
arrays.

gasche reviewed (approved):
I have read the new implementation and believe that it is correct and efficient. Do we have tests?

shindere commented:
Gabriel Scherer (2018/12/08 04:32 -0800):
> Do we have tests?

Implicitly, yes, because this PR has initially been submitted to fix
tests that were failing in flat-float-array mode. Just that these tests
are run only by Inria CI's other-configs job, which is only run on
trunk at the moment.

yallop commented on runtime/array.c:
```diff
@@ -346,6 +335,25 @@ CAMLprim value caml_make_vect(value len, value init)
+  return caml_floatarray_create (len);
+#else
+  if (uninitialized_float == Val_unit){
+    caml_register_global_root (&uninitialized_float);
```
Could this be a generational root to avoid the tiny (few ns) hit of scanning it on every minor collection?

damiendoligez replied:
Done.

yallop commented on runtime/array.c:
```diff
@@ -346,6 +335,25 @@ CAMLprim value caml_make_vect(value len, value init)
 }
 
+#ifndef FLAT_FLOAT_ARRAY
+static value uninitialized_float = Val_unit;
```
`uninitialized_float` could be moved into `caml_make_float_vect` to reduce its scope, and to avoid the need for two sets of `#ifdef` conditions.

damiendoligez replied:
Done.

damiendoligez committed:
move uninitialized_float into caml_make_float_vect and make it a generational root

damiendoligez commented:
Thanks @gasche and @yallop. I'll merge after CI passes.

shindere commented:
Damien Doligez (2018/12/28 05:14 -0800):
> Merged #2183 into trunk.

Great! Thanks!!
